components/productList/FilteredProductList/attributes.ts

```diff
@@ -0,0 +1,73 @@
+import { AttributeFilterFragment, OrderDirection, ProductOrderField } from "@/saleor/api";
+
+import { FilterPill } from "./FilterPills";
+
+export interface UrlFilter {
+  slug: string;
+  values: string[];
+}
+export interface UrlSorting {
+  field: ProductOrderField;
+  direction: OrderDirection;
+}
+
+export const getPillsData = (f: UrlFilter[], attributeFiltersData: AttributeFilterFragment[]) => {
+  const pills: FilterPill[] = [];
+  for (const filter of f) {
+    const choiceAttribute = attributeFiltersData.find((a) => a.slug === filter.slug);
+    const attrName = choiceAttribute ? choiceAttribute.name : filter.slug;
+    for (const value of filter.values) {
+      let choiceName = value;
+      if (choiceAttribute) {
+        const attrChoice = choiceAttribute.choices?.edges.find((ch) => ch.node.slug === value);
+        if (attrChoice?.node.name) {
+          choiceName = attrChoice.node.name;
+        }
+      }
+      pills.push({
+        label: `${attrName}: ${choiceName}`,
+        choiceSlug: value,
+        attributeSlug: filter.slug,
+      });
+    }
+  }
+  return pills;
+};
+
+export const parseQueryAttributeFilters = (query: string): UrlFilter[] => {
```

Comment:
There's this package here that simplifies entire query parsing and serialization: https://www.npmjs.com/package/qs

Reply:
Don't think it's worth adding to the bundle size (https://bundlephobia.com/package/qs@6.10.3). This solution is inspired by what has been used in the old storefront.

BTW pkg we are using for URL state has some serialization options, but URL will not be as easy to read as with this custom method.

Reply:
Fair enough, that's a pretty bloated package 👀